core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
```

Comment:
Lets set away streams at all. We have kafka client and we assign a topic/partition to it. The question is - from what partition client should read? There are two solutions: get offset from kafka builtin offset storage or get it from somewhere (like database or so). 

We are interested in builtin mechanism here. So, we want to get offset from kafka and kafka have no idea about this consumer. Then you use predefined behavior: reading from start or from end. Next, when you got message you have to signal kafka builtin offset storage that you done with message. This is what we call here offset committing. So committing offset is just command to kafka to remember your offset. In kafka it is (consumerId, topic, partition) -> offset.

To achieve at-most-once I have to commit offset immediately after fetching it. But commit is asynchronous in kafka client API, then we have to wait confirmation.

Pseudo code will be the same:
// here we have no offset setup at all
fetch message
commit offset and wait for commit confirmation // after confirmation I am sure that after restarting kafka client will automatically set offset to commited.
perform A => B
publish to next topic